primitives/src/signatures/bls.rs

```diff
@@ -1,157 +1,316 @@
+//! assert!(BLSSignatureScheme::verify(&pp, &pk1, &msg, &sig).is_ok());
+//!
+//! let msg = "I agree to the Terms and Conditions.";
+//! let sig = BLSSignatureScheme::sign(&pp, &sk2, &msg, &mut OsRng)?;
+//! assert!(BLSSignatureScheme::verify(&pp, &pk2, &msg, &sig).is_ok());
+//!
+//! # Ok::<(), Box<dyn std::error::Error>>(())
+//! ```
+//!
+//! [zeroize]: https://github.com/RustCrypto/utils/tree/master/zeroize
 
 use super::SignatureScheme;
 use crate::{
     constants::{
-        BLS_SIG_KEY_SIZE, BLS_SIG_SIGNATURE_SIZE, BLS_SIG_VERKEY_SIZE, CS_ID_BLS_SIG_NAIVE,
+        BLS_SIG_COMPRESSED_PK_SIZE, BLS_SIG_COMPRESSED_SIGNATURE_SIZE, BLS_SIG_PK_SIZE,
+        BLS_SIG_SIGNATURE_SIZE, BLS_SIG_SK_SIZE, CS_ID_BLS_MIN_SIG,
     },
     errors::PrimitivesError,
 };
 
-use ark_serialize::{CanonicalDeserialize, CanonicalSerialize, SerializationError};
+use ark_serialize::*;
 use ark_std::{
     convert::TryInto,
     format,
+    ops::{Deref, DerefMut},
     rand::{CryptoRng, RngCore},
+    vec::Vec,
 };
-
-use blst::BLST_ERROR;
+use blst::{min_sig::*, BLST_ERROR};
 use espresso_systems_common::jellyfish::tag;
 use tagged_base64::tagged;
+use zeroize::{Zeroize, Zeroizing};
 
-pub use blst::min_sig::{PublicKey, SecretKey, Signature};
-use zeroize::Zeroize;
-
-/// Newtype wrapper for a BLS Signing Key.
 #[tagged(tag::BLS_SIGNING_KEY)]
-#[derive(Clone, Debug, Zeroize)]
+#[derive(Clone, Debug, Default, Zeroize)]
+#[zeroize(drop)]
+/// A BLS Secret Key (Signing Key).
 pub struct BLSSignKey(SecretKey);
 
-impl core::ops::Deref for BLSSignKey {
+impl Deref for BLSSignKey {
     type Target = SecretKey;
-
     fn deref(&self) -> &Self::Target {
         &self.0
     }
 }
 
 impl CanonicalSerialize for BLSSignKey {
-    fn serialized_size(&self) -> usize {
-        BLS_SIG_KEY_SIZE
+    fn serialize<W: Write>(&self, writer: W) -> Result<(), SerializationError> {
+        CanonicalSerialize::serialize(&self.to_bytes()[..], writer)
     }
 
-    fn serialize<W: ark_serialize::Write>(&self, writer: W) -> Result<(), SerializationError> {
-        let bytes = &self.0.serialize();
-        CanonicalSerialize::serialize(bytes.as_ref(), writer)
+    fn serialized_size(&self) -> usize {
+        BLS_SIG_SK_SIZE
     }
 }
 
 impl CanonicalDeserialize for BLSSignKey {
-    fn deserialize<R: ark_serialize::Read>(mut reader: R) -> Result<Self, SerializationError> {
+    fn deserialize<R: Read>(mut reader: R) -> Result<Self, SerializationError> {
         let len = <usize as ark_serialize::CanonicalDeserialize>::deserialize(&mut reader)?;
-        if len != BLS_SIG_KEY_SIZE {
+        if len != BLS_SIG_SK_SIZE {
             return Err(SerializationError::InvalidData);
         }
 
-        let mut key = [0u8; BLS_SIG_KEY_SIZE];
-        reader.read_exact(&mut key)?;
-        SecretKey::deserialize(&key)
+        let mut sk_bytes = [0u8; BLS_SIG_SK_SIZE];
+        reader.read_exact(&mut sk_bytes)?;
+        SecretKey::from_bytes(&sk_bytes)
```

Comment:
`from_bytes` is just a wrapper on `deserialize`

Reply:
yes, I know. Okay, I will change to your impl.